internal/server/iito.go

```diff
@@ -48,56 +48,57 @@ func (s *SOMASServer) runGiftSession() error {
-	var err error
+
+// GetGiftOffers collects all responses from clients to their requests in a map
+func (s *SOMASServer) getGiftOffers(totalRequests map[shared.ClientID]shared.GiftRequestDict) (map[shared.ClientID]shared.GiftOfferDict, error) {
+	totalOffers := map[shared.ClientID]shared.GiftOfferDict{}
+	// Loop over each team
 	for id, client := range s.clientMap {
-		giftOfferDict[id], err = client.OfferGifts(giftRequestDict)
+		// Gather all the requests made to this team
+		requestsToThisTeam := shared.GiftRequestDict{}
+		for fromTeam, indivRequests := range totalRequests {
+			requestsToThisTeam[fromTeam] = indivRequests[id]
+		}
+		offer, err := client.GetGiftOffers(requestsToThisTeam)
 		if err != nil {
-			return giftOfferDict, err
+			totalOffers[id] = offer
 		}
 	}
-	return giftOfferDict, nil
+	return totalOffers, nil
 }
-func (s *SOMASServer) getGiftAcceptance(giftOffersDict map[shared.ClientID]shared.GiftDict) (map[shared.ClientID]shared.GiftInfoDict, error) {
-	acceptedGifts := map[shared.ClientID]shared.GiftInfoDict{}
-	var err error
 
-	receivedByClientDict := make(map[shared.ClientID]shared.GiftDict)
-
-	//puts all the gifts received by a certain client accesible by the id of that client
-	for idsend := range giftOffersDict {
-		for idto := range giftOffersDict {
-			if idsend == idto {
-				continue
-			}
-			if receivedByClientDict[idsend] == nil {
-				receivedByClientDict[idsend] = make(shared.GiftDict)
-			}
-			receivedByClientDict[idsend][idto] = giftOffersDict[idto][idsend]
-		}
-	}
+func (s *SOMASServer) getGiftResponses(totalOffers map[shared.ClientID]shared.GiftOfferDict) (map[shared.ClientID]shared.GiftResponseDict, error) {
+	totalResponses := map[shared.ClientID]shared.GiftResponseDict{}
 
 	for id, client := range s.clientMap {
-		acceptedGifts[id], err = client.AcceptGifts(receivedByClientDict[id])
+		offersToThisTeam := shared.GiftOfferDict{}
+		for fromTeam, indivOffers := range totalOffers {
+			offersToThisTeam[fromTeam] = indivOffers[id]
+		}
+		response, err := client.GetGiftResponses(offersToThisTeam)
 		if err != nil {
-			return acceptedGifts, err
+			totalResponses[id] = response
 		}
 	}
-	return acceptedGifts, nil
+	return totalResponses, nil
 }
 
-func (s *SOMASServer) distributeGiftHistory(acceptedGifts map[shared.ClientID]shared.GiftInfoDict) error {
-	//Process acceptedGifts
-	for _, client := range s.clientMap {
-		err := client.UpdateGiftInfo(acceptedGifts)
+// distributeGiftHistory collates all responses to a single client and calls that client to receive its responses.
+func (s *SOMASServer) distributeGiftHistory(totalResponses map[shared.ClientID]shared.GiftResponseDict) error {
+	for id, client := range s.clientMap {
+		responsesToThisTeam := shared.GiftResponseDict{}
+		for fromTeam, indivResponses := range totalResponses {
+			responsesToThisTeam[fromTeam] = indivResponses[id]
+		}
+		err := client.UpdateGiftInfo(responsesToThisTeam)
```

Comment:
Should clients be able to send offers to dead clients?

Reply:
Nope! Again, awaiting the exposure of a getNonDeadClients equivalent from the server on another PR.

Reply:
Sounds good.